cli/src/main/kotlin/commands/UploadResultToSw360Command.kt

```diff
@@ -0,0 +1,189 @@
+    epilog = "EXPERIMENTAL: The command is still in development and usage will likely change in the near future. The " +
+            "command expects that a Sw360Storage for the scanner is configured."
+) {
+    private val ortFile by option(
+        "--ort-file", "-i",
+        help = "The ORT result file to read as input."
+    ).convert { it.expandTilde() }
+        .file(mustExist = true, canBeFile = true, canBeDir = false, mustBeWritable = false, mustBeReadable = true)
+        .convert { it.absoluteFile.normalize() }
+        .required()
+        .inputGroup()
+
+    private val sw360JsonMapper = jsonMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    private val globalOptionsForSubcommands by requireObject<GlobalOptions>()
+
+    override fun run() {
+        val (ortResult, duration) = measureTimedValue { ortFile.readValue<OrtResult>() }
+
+        log.perf {
+            "Read ORT result from '${ortFile.name}' (${ortFile.formatSizeInMib}) in ${duration.inMilliseconds}ms."
+        }
+
+        val sw360Config = globalOptionsForSubcommands.config.scanner?.storages?.values
+            ?.filterIsInstance<Sw360StorageConfiguration>()?.singleOrNull()
+
+        requireNotNull(sw360Config) {
+            "No SW360 storage is configured for the scanner."
+        }
+
+        val httpClientConfig = HttpClientConfig
+            .basicConfig()
+            .withObjectMapper(sw360JsonMapper)
+        val newHttpClient = HttpClientFactoryImpl().newHttpClient(httpClientConfig)
+
+        requireNotNull(newHttpClient) {
+            "Could not create new HTTP client instance."
+        }
+
+        val sw360ClientConfig = SW360ClientConfig.createConfig(
+            sw360Config.restUrl,
+            sw360Config.authUrl,
+            sw360Config.username,
+            sw360Config.password,
+            sw360Config.clientId,
+            sw360Config.clientPassword,
+            newHttpClient,
+            sw360JsonMapper
+        )
+        val sw360Connection = SW360ConnectionFactory().newConnection(sw360ClientConfig)
+
+        requireNotNull(sw360Connection) {
+            "Could not establish a SW360 connection with the provided SW360 client configuration."
+        }
+
+        val sw360ReleaseClient = sw360Connection.releaseAdapter
+        val sw360ProjectClient = sw360Connection.projectAdapter
+
+        val projectWithPackages = ortResult.getProjects(omitExcluded = true).map { project ->
+            project to ortResult.getPackages().filter { (pkg, _) -> pkg.id in project.collectDependencies() }
+                .map { (pkg, _) -> pkg }
+        }.toMap()
+
+        projectWithPackages.forEach { (project, pkgList) ->
+            val linkedReleases = pkgList.mapNotNull { pkg ->
+                val name = listOfNotNull(pkg.id.namespace, pkg.id.name).joinToString("/")
+                sw360ReleaseClient.getSparseReleaseByNameAndVersion(name, pkg.id.version)
+                    .takeIf { it.isPresent }
```

Comment:
You use this pretty verbose idiom to deal with Optional at multiple places. I think this could be simplified. For instance for the combination of getSparseReleaseByNameAndVersion() and enrichtSparseRelease(), you could probably use `Optional.flatMap()`. On other places, `Optional.orElse(null)` is shorter than a `takeIf` followed by a `let { x.get() } `.

Reply:
I wanted to go the "kotlinish" way and avoid using Java's `Optional` methods. Do you think it's better to use them?

Reply:
I would say, use everything what makes the code elegant and easy to read. Probably `Optional` does not play very well with Kotlin as here nullable types are used to express the same concept. But from my gut feeling, the repeated combination of `takeIf()` and `let()` is not ideal.